python/lsst/ip/diffim/subtractImages.py

```diff
@@ -617,6 +605,232 @@ def _sourceSelector(self, sources):
+        " with those in finalizedPsfApCorrCatalog.",
+        dtype=bool,
+        default=False,
+    )
+    detectionThreshold = lsst.pex.config.Field(
+        dtype=float,
+        default=10,
+        doc="Minimum signal to noise ration of detected sources "
+        "to use for calculating the PSF matching kernel."
+    )
+    badSourceFlags = lsst.pex.config.ListField(
+        dtype=str,
+        doc="Flags that, if set, the associated source should not "
+        "be used to determine the PSF matching kernel.",
+        default=("sky_source", "slot_Centroid_flag",
+                 "slot_ApFlux_flag", "slot_PsfFlux_flag", ),
+    )
+
+    def setDefaults(self):
+        self.makeKernel.kernel.name = "AL"
+        self.makeKernel.kernel.active.fitForBackground = True
+        self.makeKernel.kernel.active.spatialKernelOrder = 1
+        self.makeKernel.kernel.active.spatialBgOrder = 2
+
+
+class AlardLuptonPreconvolveSubtractTask(AlardLuptonSubtractTask):
+    """Subtract a template from a science image, convolving the science image
+    before computing the kernel, and also convolving the template before
+    subtraction.
+    """
+    ConfigClass = AlardLuptonPreconvolveSubtractConfig
+    _DefaultName = "alardLuptonPreconvolveSubtract"
+
+    def run(self, template, science, sources, finalizedPsfApCorrCatalog=None):
+        """Preconvolve the science image with its own PSF,
+        convolve the template image with a PSF-matching kernel and subtract
+        from the preconvolved science image.
+
+        Parameters
+        ----------
+        template : `lsst.afw.image.ExposureF`
+            The template image, which has previously been warped to
+            the science image. The template bbox will be padded by a few pixels
+            compared to the science bbox.
+        science : `lsst.afw.image.ExposureF`
+            The science exposure.
+        sources : `lsst.afw.table.SourceCatalog`
+            Identified sources on the science exposure. This catalog is used to
+            select sources in order to perform the AL PSF matching on stamp
+            images around them.
+
+        Returns
+        -------
+        results : `lsst.pipe.base.Struct`
+            ``scoreExposure`` : `lsst.afw.image.ExposureF`
+                The image difference.
+            ``matchedTemplate`` : `lsst.afw.image.ExposureF`
+                The warped and PSF-matched template exposure.
+        """
+
+        selectSources = self.prepareInputs(template, science, sources,
+                                           finalizedPsfApCorrCatalog=finalizedPsfApCorrCatalog)
+
+        subtractResults = self.runConvolveTemplate(template, science, selectSources)
+
+        return subtractResults
+
+    def runConvolveTemplate(self, template, science, selectSources):
+        """Convolve the template image with a PSF-matching kernel and subtract
+        from the science image.
```

Comment:
I think this docstring is incorrect, and the `Returns` below certainly is.

I also wonder if this shouldn't be called something different, since it's not really doing the same thing as `runConvolveTemplate` in the other class. `runPreconvolve`?